exec/java-exec/src/test/java/org/apache/drill/test/ProfileParser.java

```diff
@@ -138,9 +414,208 @@ public long getMetric(int id) {
+  {
+    public void visit(OpDefInfo root) {
+      visit(root, 0);
+    }
+    public void visit(OpDefInfo node, int indent) {
+      visitOp( node, indent );
+      if (node.children.isEmpty()) {
+        return;
+      }
+      if ( node.children.size() == 1) {
+        visit(node.children.get(0), indent);
+        return;
+      }
+      indent++;
+      int i = 0;
+      for (OpDefInfo child : node.children) {
+        visitSubtree(node, i++, indent);
+        visit(child, indent+1);
+      }
+    }
+
+    protected void visitOp(OpDefInfo node, int indent) {
+    }
+
+    protected void visitSubtree(OpDefInfo node, int i, int indent) {
+    }
+
+    public String indentString(int indent, String pad) {
+      StringBuilder buf = new StringBuilder();
+      for (int i = 0; i < indent; i++) {
+        buf.append( pad );
+      }
+      return buf.toString();
+    }
+
+    public String indentString(int indent) {
+      return indentString(indent, "  ");
+    }
+
+    public String subtreeLabel(OpDefInfo node, int branch) {
+      if (node.name.equals("HashJoin")) {
+        return (branch == 0) ? "Probe" : "Build";
+      } else {
+        return "Input " + (branch + 1);
+      }
+    }
+  }
+
+  /**
+   * Print the operator tree for analysis.
+   */
+
+  public static class TreePrinter extends TreeVisitor
+  {
+    @Override
+    protected void visitOp(OpDefInfo node, int indent) {
+      System.out.print( indentString(indent) );
+      System.out.println( node.toString() );
+    }
+
+    @Override
+    protected void visitSubtree(OpDefInfo node, int i, int indent) {
+      System.out.print( indentString(indent) );
+      System.out.println(subtreeLabel(node, i));
+    }
+  }
+
+  /**
+   * Print out the tree showing a comparison of estimated vs.
+   * actual costs.
+   */
+
+  public static class CostPrinter extends TreeVisitor
+  {
+    @Override
+    protected void visitOp(OpDefInfo node, int indentLevel) {
+      System.out.print(String.format("%02d-%02d ", node.majorId, node.stepId));
+      String indent = indentString(indentLevel, ". ");
+      System.out.print( indent + node.name );
+      if (node.opName != null) {
+        System.out.print( " (" + node.opName + ")" );
+      }
+      System.out.println( );
+      indent = indentString(15);
+      System.out.print( indent );
+      System.out.println(String.format("  Estimate: %,15.0f rows, %,7.0f MB",
+                         node.estRows, node.estMemoryCost / 1024 / 1024) );
+      System.out.print( indent );
+      System.out.println(String.format("  Actual:   %,15d rows, %,7d MB",
+                         node.actualRows, node.actualMemory / 1024 / 1024));
+    }
+
+    @Override
+    protected void visitSubtree(OpDefInfo node, int i, int indent) {
+      System.out.print( indentString(indent) + "      " );
+      System.out.println(subtreeLabel(node, i));
+    }
+  }
+
+  public Map<Integer,OperatorProfile> getOpInfo( ) {
     Map<Integer,String> ops = getOperators( );
```

Comment:
Added some Javadoc. This whole class is under constant construction, so I've been lazy about adding comments for stuff that will likely change later...